File: apps/node/tests/stack_queue.test.ts
/**
 * 第4章 スタックとキュー — テスト
 */
import { FixedStack, FixedQueue } from '../src/algorithm/stack_queue';

describe('固定長スタック', () => {
  let stack: FixedStack<number>;

  beforeEach(() => {
    stack = new FixedStack(64);
  });

  test('初期状態: 空で満杯でない', () => {
    expect(stack.isEmpty()).toBe(true);
    expect(stack.isFull()).toBe(false);
  });

  test('push して peek', () => {
    stack.push(1);
    expect(stack.peek()).toBe(1);
  });

  test('push して pop', () => {
    stack.push(1);
    expect(stack.pop()).toBe(1);
  });

  test('LIFO 順序', () => {
    stack.push(1);
    stack.push(2);
    stack.push(3);
    expect(stack.pop()).toBe(3);
    expect(stack.pop()).toBe(2);
    expect(stack.pop()).toBe(1);
  });

  test('満杯判定', () => {
    const small = new FixedStack<number>(3);
    small.push(1);
    small.push(2);
    small.push(3);
    expect(small.isFull()).toBe(true);
  });

  test('空のスタックから pop すると例外', () => {
    expect(() => stack.pop()).toThrow('Stack is empty');
  });

  test('満杯のスタックに push すると例外', () => {
    const small = new FixedStack<number>(2);
    small.push(1);
    small.push(2);
    expect(() => small.push(3)).toThrow('Stack is full');
  });

  test('空のスタックを peek すると例外', () => {
    expect(() => stack.peek()).toThrow('Stack is empty');
  });

  test('find: 底から検索', () => {
    stack.push(10);
    stack.push(20);
    stack.push(30);
    expect(stack.find(20)).toBe(1);
  });

  test('find: 存在しない場合は -1', () => {
    stack.push(10);
    expect(stack.find(99)).toBe(-1);
  });

  test('count: 同じ値の個数', () => {
    stack.push(5);
    stack.push(5);
    stack.push(10);
    expect(stack.count(5)).toBe(2);
    expect(stack.count(10)).toBe(1);
    expect(stack.count(99)).toBe(0);
  });

  test('contains: 包含判定', () => {
    stack.push(42);
    expect(stack.contains(42)).toBe(true);
    expect(stack.contains(0)).toBe(false);
  });

  test('clear: スタックを空にする', () => {
    stack.push(1);
    stack.push(2);
    stack.clear();
    expect(stack.isEmpty()).toBe(true);
  });

  test('size: 要素数', () => {
    stack.push(1);
    stack.push(2);
    expect(stack.size()).toBe(2);
  });

  test('capacity: 容量', () => {
    expect(stack.capacity).toBe(64);
  });
});

describe('固定長キュー（リングバッファ）', () => {
  let queue: FixedQueue<number>;

  beforeEach(() => {
    queue = new FixedQueue(64);
  });

  test('初期状態: 空で満杯でない', () => {
    expect(queue.isEmpty()).toBe(true);
    expect(queue.isFull()).toBe(false);
  });

  test('enqueue して dequeue', () => {
    queue.enqueue(1);
    expect(queue.dequeue()).toBe(1);
  });

  test('FIFO 順序', () => {
    queue.enqueue(1);
    queue.enqueue(2);
    queue.enqueue(3);
    expect(queue.dequeue()).toBe(1);
    expect(queue.dequeue()).toBe(2);
    expect(queue.dequeue()).toBe(3);
  });

  test('peek: 先頭要素を参照', () => {
    queue.enqueue(10);
    queue.enqueue(20);
    expect(queue.peek()).toBe(10);
  });

  test('満杯判定', () => {
    const small = new FixedQueue<number>(3);
    small.enqueue(1);
    small.enqueue(2);
    small.enqueue(3);
    expect(small.isFull()).toBe(true);
  });

  test('空のキューから dequeue すると例外', () => {
    expect(() => queue.dequeue()).toThrow('Queue is empty');
  });

  test('満杯のキューに enqueue すると例外', () => {
    const small = new FixedQueue<number>(2);
    small.enqueue(1);
    small.enqueue(2);
    expect(() => small.enqueue(3)).toThrow('Queue is full');
  });

  test('空のキューを peek すると例外', () => {
    expect(() => queue.peek()).toThrow('Queue is empty');
  });

  test('find: 先頭からの検索', () => {
    queue.enqueue(10);
    queue.enqueue(20);
    queue.enqueue(30);
    expect(queue.find(10)).toBe(0);
  });

  test('find: 存在しない場合は -1', () => {
    queue.enqueue(10);
    expect(queue.find(99)).toBe(-1);
  });

  test('count: 同じ値の個数', () => {
    queue.enqueue(5);
    queue.enqueue(5);
    queue.enqueue(10);
    expect(queue.count(5)).toBe(2);
  });

  test('contains: 包含判定', () => {
    queue.enqueue(42);
    expect(queue.contains(42)).toBe(true);
    expect(queue.contains(0)).toBe(false);
  });

  test('clear: キューを空にする', () => {
    queue.enqueue(1);
    queue.enqueue(2);
    queue.clear();
    expect(queue.isEmpty()).toBe(true);
  });

  test('size: 要素数', () => {
    queue.enqueue(1);
    queue.enqueue(2);
    expect(queue.size()).toBe(2);
  });

  test('capacity: 容量', () => {
    expect(queue.capacity).toBe(64);
  });

  test('リングバッファの折り返し動作', () => {
    const small = new FixedQueue<number>(3);
    small.enqueue(1);
    small.enqueue(2);
    small.enqueue(3);
    small.dequeue(); // 1 を取り出す
    small.enqueue(4); // 空き位置（先頭）に追加
    expect(small.dequeue()).toBe(2);
    expect(small.dequeue()).toBe(3);
    expect(small.dequeue()).toBe(4);
  });
});

File: apps/go/chapter04/stack_queue_test.go
package chapter04_test

import (
	"testing"

	"github.com/k2works/getting-started-algorithm/go/chapter04"
)

func TestStack(t *testing.T) {
	s := chapter04.NewStack(5)

	if !s.IsEmpty() {
		t.Error("new stack should be empty")
	}

	s.Push(1)
	s.Push(2)
	s.Push(3)

	val, err := s.Peek()
	if err != nil || val != 3 {
		t.Errorf("Peek() = %v, %v; want 3, nil", val, err)
	}

	val, err = s.Pop()
	if err != nil || val != 3 {
		t.Errorf("Pop() = %v, %v; want 3, nil", val, err)
	}

	if s.Len() != 2 {
		t.Errorf("Len() = %d; want 2", s.Len())
	}

	if !s.Contains(2) {
		t.Error("Contains(2) should be true")
	}
}

func TestStackEmpty(t *testing.T) {
	s := chapter04.NewStack(3)
	_, err := s.Pop()
	if err == nil {
		t.Error("Pop on empty stack should return error")
	}
}

func TestStackFull(t *testing.T) {
	s := chapter04.NewStack(2)
	s.Push(1)
	s.Push(2)
	err := s.Push(3)
	if err == nil {
		t.Error("Push on full stack should return error")
	}
}

func TestQueue(t *testing.T) {
	q := chapter04.NewQueue(5)

	if !q.IsEmpty() {
		t.Error("new queue should be empty")
	}

	q.Enqueue(1)
	q.Enqueue(2)
	q.Enqueue(3)

	val, err := q.Dequeue()
	if err != nil || val != 1 {
		t.Errorf("Dequeue() = %v, %v; want 1, nil", val, err)
	}

	if q.Len() != 2 {
		t.Errorf("Len() = %d; want 2", q.Len())
	}
}

func TestQueueEmpty(t *testing.T) {
	q := chapter04.NewQueue(3)
	_, err := q.Dequeue()
	if err == nil {
		t.Error("Dequeue on empty queue should return error")
	}
}

func TestStackFindCountClear(t *testing.T) {
	s := chapter04.NewStack(5)
	s.Push(1)
	s.Push(2)
	s.Push(1)

	if s.Find(2) == -1 {
		t.Error("Find(2) should find 2")
	}
	if s.Find(99) != -1 {
		t.Error("Find(99) should return -1")
	}
	if s.Count(1) != 2 {
		t.Errorf("Count(1) = %d; want 2", s.Count(1))
	}
	s.Clear()
	if s.Len() != 0 {
		t.Errorf("after Clear Len() = %d; want 0", s.Len())
	}
}

func TestQueuePeekFindCountClear(t *testing.T) {
	q := chapter04.NewQueue(5)
	q.Enqueue(1)
	q.Enqueue(2)
	q.Enqueue(1)

	val, err := q.Peek()
	if err != nil || val != 1 {
		t.Errorf("Peek() = %v, %v; want 1, nil", val, err)
	}
	if q.Find(2) != 1 {
		t.Errorf("Find(2) = %d; want 1", q.Find(2))
	}
	if q.Find(99) != -1 {
		t.Error("Find(99) should return -1")
	}
	if q.Count(1) != 2 {
		t.Errorf("Count(1) = %d; want 2", q.Count(1))
	}
	q.Clear()
	if q.Len() != 0 {
		t.Errorf("after Clear Len() = %d; want 0", q.Len())
	}
}
